### ISSUE TITLE
Partial indexes work wrong or I use it wrong?

### ISSUE THREAD

**Issue by iswarezwp** [yugabyte/yugabyte-db#5510]
```sql
create table shipments(id int, delivery_status text, address text, delivery_date date);
create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';

explain select * from shipments where delivery_status='a';
explain select * from shipments where delivery_status='delivered';
explain select * from shipments where address='1';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);

-- query data 
select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date
----+-----------------+---------+---------------
  3 | not delivered   | 3       |
  2 | not delivered   | 2       |
(2 rows)

-- update one record from 'not delivered' to 'delivered'
update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;

-- query data again
select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | delivered       | 2       | 2020-12-12
(2 rows)
```

The last query should not return the record where id=2 ??

**Comment by jaki** [comment]
Hi, @iswarezwp.  I was not able to reproduce.  I got

```
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)
```

Can you provide outputs of `\di`, `\d+ shipments`, `explain select * from shipments where delivery_status != 'delivered';`?

**Comment by iswarezwp** [comment]
@jaki  It's very strange. I updated the test steps, execute some `explain` before `insert`，you can try again.

**Comment by jaki** [comment]
@iswarezwp, I wanted to know the output of those three commands instead.  After that, can you try reproducing on a fresh cluster?

Providing Yugabyte version and platform can also help.

**Comment by iswarezwp** [comment]
```sql
yugabyte=# \di
                     List of relations
 Schema |       Name        | Type  |  Owner   |   Table   
--------+-------------------+-------+----------+-----------
 public | shipment_delivery | index | yugabyte | shipments
(1 row)

yugabyte=# \d+ shipments
                                      Table "public.shipments"
     Column      |  Type   | Collation | Nullable | Default | Storage  | Stats target | Description 
-----------------+---------+-----------+----------+---------+----------+--------------+-------------
 id              | integer |           |          |         | plain    |              | 
 delivery_status | text    |           |          |         | extended |              | 
 address         | text    |           |          |         | extended |              | 
 delivery_date   | date    |           |          |         | plain    |              | 
Indexes:
    "shipment_delivery" lsm (address HASH, delivery_date) WHERE delivery_status <> 'delivered'::text

yugabyte=# explain select * from shipments where delivery_status != 'delivered';
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Index Scan using shipment_delivery on shipments  (cost=0.00..102.00 rows=1000 width=72)
(1 row)
```

**Comment by iswarezwp** [comment]
@jaki I reproduced it twice on fresh cluster，but the third time it's OK, I don't know why.

**Comment by jaki** [comment]
Thanks, @iswarezwp.  For the bad cases, let's see what data is actually in the index.

1. Flush the in-memory data to sst files: `bin/yb-admin --master_addresses 127.0.0.1 flush_table ysql.yugabyte shipment_delivery`
1. Inspect the sst files: `for sst_file in $(find path/to/yugabyte-data -name \*.sst | grep $(./bin/yb-admin --master_addresses 127.0.0.1 list_tables include_table_id | grep yugabyte.shipment_delivery | awk '{print$2}')); do ./bin/sst_dump --file="$sst_file" --output_hex --command=scan; done`

I made some assumptions, so adjust the commands if needed.

**Comment by iswarezwp** [comment]
I reproduced it again，this time I run the following sql:
```sql
ysqlsh (11.2-YB-2.2.0.0-b0)
Type "help" for help.

yugabyte=# create table shipments(id int, delivery_status text, address text, delivery_date date);
plain select * from shipments where delivery_status='delivered';
explain select * from shipments where address='1';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);
CREATE TABLE
yugabyte=# create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';
CREATE INDEX
yugabyte=# 
yugabyte=# explain select * from shipments where delivery_status='a';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'a'::text)
(2 rows)

yugabyte=# explain select * from shipments where delivery_status='delivered';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'delivered'::text)
(2 rows)

yugabyte=# explain select * from shipments where address='1';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (address = '1'::text)
(2 rows)

yugabyte=# 
yugabyte=# insert into shipments values (1, 'delivered', '1', '2011-11-11');
INSERT 0 1
yugabyte=# insert into shipments values (2, 'not delivered', '2', null);
INSERT 0 1
yugabyte=# insert into shipments values (3, 'not delivered', '3', null);
INSERT 0 1
yugabyte=# 
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | not delivered   | 2       | 
(2 rows)

yugabyte=# update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
UPDATE 1
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)

yugabyte=# drop table shipments ;
DROP TABLE
yugabyte=# create table shipments(id int, delivery_status text, address text, delivery_date date);
CREATE TABLE
yugabyte=# create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';
CREATE INDEX
yugabyte=# insert into shipments values (1, 'delivered', '1', '2011-11-11');
INSERT 0 1
yugabyte=# insert into shipments values (2, 'not delivered', '2', null);
INSERT 0 1
yugabyte=# insert into shipments values (3, 'not delivered', '3', null);
INSERT 0 1
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | not delivered   | 2       | 
(2 rows)

yugabyte=# update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
UPDATE 1
yugabyte=# 
yugabyte=# -- query data again
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)

yugabyte=# -- query data again
select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)

yugabyte=# -- query data again
select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)

yugabyte=# -- query data again
select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
(1 row)

yugabyte=# explain select * from shipments where delivery_status='a';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'a'::text)
(2 rows)

yugabyte=# explain select * from shipments where delivery_status='delivered';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'delivered'::text)
(2 rows)

yugabyte=# explain select * from shipments where address='1';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (address = '1'::text)
(2 rows)

yugabyte=# insert into shipments values (1, 'delivered', '1', '2011-11-11');
INSERT 0 1
yugabyte=# insert into shipments values (2, 'not delivered', '2', null);
INSERT 0 1
yugabyte=# insert into shipments values (3, 'not delivered', '3', null);
INSERT 0 1
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  3 | not delivered   | 3       | 
  2 | not delivered   | 2       | 
(3 rows)

yugabyte=# update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
UPDATE 2
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  3 | not delivered   | 3       | 
(2 rows)

yugabyte=# CREATE TABLE products(id int PRIMARY KEY, name text, code text);
CREATE TABLE
yugabyte=# CREATE UNIQUE INDEX ON products(code);
CREATE INDEX
yugabyte=# CREATE INDEX products_name ON products(name ASC);
CREATE INDEX
yugabyte=# CREATE INDEX products_name_code ON products(name) INCLUDE (code);
CREATE INDEX
yugabyte=# CREATE TABLE employees (id int PRIMARY KEY, first_name TEXT, last_name TEXT) SPLIT INTO 10 TABLETS;
CREATE TABLE
yugabyte=# CREATE INDEX ON employees(first_name, last_name) SPLIT INTO 10 TABLETS;
CREATE INDEX
yugabyte=# create table shipments(id int, delivery_status text, address text, delivery_date date);
ERROR:  relation "shipments" already exists
yugabyte=# drop table shipments 
yugabyte-# ;
DROP TABLE
yugabyte=# create table shipments(id int, delivery_status text, address text, delivery_date date);
CREATE TABLE
yugabyte=# create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';
CREATE INDEX
yugabyte=# explain select * from shipments where delivery_status='a';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'a'::text)
(2 rows)

yugabyte=# explain select * from shipments where delivery_status='delivered';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (delivery_status = 'delivered'::text)
(2 rows)

yugabyte=# explain select * from shipments where address='1';
                            QUERY PLAN                             
-------------------------------------------------------------------
 Foreign Scan on shipments  (cost=0.00..102.50 rows=1000 width=72)
   Filter: (address = '1'::text)
(2 rows)

yugabyte=# 
yugabyte=# insert into shipments values (1, 'delivered', '1', '2011-11-11');
INSERT 0 1
yugabyte=# insert into shipments values (2, 'not delivered', '2', null);
INSERT 0 1
yugabyte=# insert into shipments values (3, 'not delivered', '3', null);
INSERT 0 1
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | not delivered   | 2       | 
(2 rows)

yugabyte=# update shipments set delivery_status = 'delivered' where id = 2;
UPDATE 1
yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | delivered       | 2       | 
(2 rows)

yugabyte=# select * from shipments where delivery_status != 'delivered';
 id | delivery_status | address | delivery_date 
----+-----------------+---------+---------------
  3 | not delivered   | 3       | 
  2 | delivered       | 2       | 
(2 rows)
```

I have 3 nodes, the sst_dump on every node is:
**node1**
```
# for sst_file in $(find /mnt/sctest -name \*.sst | grep $(./bin/yb-admin --master_addresses 192.168.80.161 list_tables include_table_id | grep yugabyte.shipment_delivery | awk '{print$2}')); do ./bin/sst_dump --file="$sst_file" --output_hex --command=scan; done
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-cd3452dcb9544f38bef78b961889722a/000010.sst
Sst file format: block-based
'47901253320000217C5347797F53425548FE2B564A3DB90998286A17A77F0001000121210000214A80238001A6662BA3DCCC804A' @ 1125899906842625: 1 => 238001A6662BA400A47B4A24
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-181a90deb90f4379b00451f1ef32da87/000010.sst
Sst file format: block-based
'473DBC53330000217C53478C9B535B11C19E8C9546EDA14BBC52725D3AA50001000121210000214A80238001A6662BA36159804A' @ 1125899906842625: 1 => 238001A6662BA3D21D7E4A24

```
**node2**
```
# for sst_file in $(find /mnt/sctest -name \*.sst | grep $(./bin/yb-admin --master_addresses 192.168.80.161 list_tables include_table_id | grep yugabyte.shipment_delivery | awk '{print$2}')); do ./bin/sst_dump --file="$sst_file" --output_hex --command=scan; done
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-181a90deb90f4379b00451f1ef32da87/000010.sst
Sst file format: block-based
'473DBC53330000217C53478C9B535B11C19E8C9546EDA14BBC52725D3AA50001000121210000214A80238001A6662BA36159804A' @ 1125899906842625: 1 => 238001A6662BA3D21D7E4A24
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-cd3452dcb9544f38bef78b961889722a/000010.sst
Sst file format: block-based
'47901253320000217C5347797F53425548FE2B564A3DB90998286A17A77F0001000121210000214A80238001A6662BA3DCCC804A' @ 1125899906842625: 1 => 238001A6662BA400A47B4A24
```

**node3**
```
# for sst_file in $(find /mnt/sctest -name \*.sst | grep $(./bin/yb-admin --master_addresses 192.168.80.161 list_tables include_table_id | grep yugabyte.shipment_delivery | awk '{print$2}')); do ./bin/sst_dump --file="$sst_file" --output_hex --command=scan; done
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-181a90deb90f4379b00451f1ef32da87/000010.sst
Sst file format: block-based
'473DBC53330000217C53478C9B535B11C19E8C9546EDA14BBC52725D3AA50001000121210000214A80238001A6662BA36159804A' @ 1125899906842625: 1 => 238001A6662BA3D21D7E4A24
from [] to []
Process /mnt/sctest/yb-data/tserver/data/rocksdb/table-000030a9000030008000000000004019/tablet-cd3452dcb9544f38bef78b961889722a/000010.sst
Sst file format: block-based
'47901253320000217C5347797F53425548FE2B564A3DB90998286A17A77F0001000121210000214A80238001A6662BA3DCCC804A' @ 1125899906842625: 1 => 238001A6662BA400A47B4A24
```

**Comment by jaki** [comment]
Thanks for the detailed steps.  I was able to reproduce on rf 1.

Looking at the `sst_dump` output, there are two records in the index:

- `'473DBC53330000217C53478C9B535B11C19E8C9546EDA14BBC52725D3AA50001000121210000214A80238001A6662BA36159804A' @ 1125899906842625: 1 => 238001A6662BA3D21D7E4A24`
- `'473DBC53330000217C53478C9B535B11C19E8C9546EDA14BBC52725D3AA50001000121210000214A80238001A6662BA36159804A' @ 1125899906842625: 1 => 238001A6662BA3D21D7E4A24`

In the good case, you should expect to see a delete record (its value ends in ascii 'X' = hex 58, which is known as `kTombstone` in the source code).  Since this delete record is not present, there appears to be an issue on the data side rather than the query side.

**Comment by jaki** [comment]
I created a copy-pasteable version of your steps:

```
create table shipments(id int, delivery_status text, address text, delivery_date date);
create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';

explain select * from shipments where delivery_status='a';
explain select * from shipments where delivery_status='delivered';
explain select * from shipments where address='1';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);

select * from shipments where delivery_status != 'delivered';
update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
select * from shipments where delivery_status != 'delivered';

drop table shipments;
create table shipments(id int, delivery_status text, address text, delivery_date date);
create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);

select * from shipments where delivery_status != 'delivered';
update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
select * from shipments where delivery_status != 'delivered';
select * from shipments where delivery_status != 'delivered';
select * from shipments where delivery_status != 'delivered';
select * from shipments where delivery_status != 'delivered';

explain select * from shipments where delivery_status='a';
explain select * from shipments where delivery_status='delivered';
explain select * from shipments where address='1';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);

select * from shipments where delivery_status != 'delivered';
update shipments set delivery_status = 'delivered', delivery_date = '2020-12-12' where id = 2;
select * from shipments where delivery_status != 'delivered';

CREATE TABLE products(id int PRIMARY KEY, name text, code text);
CREATE UNIQUE INDEX ON products(code);
CREATE INDEX products_name ON products(name ASC);
CREATE INDEX products_name_code ON products(name) INCLUDE (code);
CREATE TABLE employees (id int PRIMARY KEY, first_name TEXT, last_name TEXT) SPLIT INTO 10 TABLETS;
CREATE INDEX ON employees(first_name, last_name) SPLIT INTO 10 TABLETS;

create table shipments(id int, delivery_status text, address text, delivery_date date);
drop table shipments;

create table shipments(id int, delivery_status text, address text, delivery_date date);
create index shipment_delivery on shipments(address, delivery_date) where delivery_status != 'delivered';

explain select * from shipments where delivery_status='a';
explain select * from shipments where delivery_status='delivered';
explain select * from shipments where address='1';

insert into shipments values (1, 'delivered', '1', '2011-11-11');
insert into shipments values (2, 'not delivered', '2', null);
insert into shipments values (3, 'not delivered', '3', null);

select * from shipments where delivery_status != 'delivered';
update shipments set delivery_status = 'delivered' where id = 2;
select * from shipments where delivery_status != 'delivered';
select * from shipments where delivery_status != 'delivered';
```

Running this on `ysqlsh` has consistently given the bad output for 2.2.0.0 but has consistently given the good output for latest (commit f48649667).  @m-iancu pointed out a related fix in commit, 679383858c3b9edd6cd66906d96de09bb1766c63, so we should look into backporting that to our 2.2 branch.

**Comment by jaki** [comment]
Tested on commit ec7da0721cfc95729dcf67ada62b247be27d5b58 of 2.2 branch: it's good.  This was probably fixed by commit 80f668054133415ff07a842bb74bf19715a466c9 (the backport I previously mentioned).  The release https://github.com/yugabyte/yugabyte-db/releases/tag/v2.2.3.0 contains it.  Sorry about the late notice.
